README.md
# Imager.js [![Build Status](https://secure.travis-ci.org/BBC-News/Imager.js.svg?branch=master)](http://travis-ci.org/BBC-News/Imager.js)

> Imager.js is an alternative solution to the issue of how to handle responsive image loading, created by developers at [BBC News](http://responsivenews.co.uk/).


## Why?

There are many responsive image solutions in the wild: `srcset`, `src-n`, `PictureFill` and so on. They are either
**verbose** or **hard to debug** (and to maintain/integrate). Some of them don't deal well with *pixel density*
and suffer from **double asset payload** (meaning you end up downloading assets unnecessarily).

We wanted something **simple**, which **works** and which is **fast** as well as network friendly (only download what you need, when you need it).

Imager implements the  [BBC Responsive News technique](http://responsivenews.co.uk/post/50092458307/images) which incorporates:

- loading any image once
- loading the most suitable sized image

## How does it work?

Imager runs through the following workflow:

1. lookup **placeholder elements**
2. replace **placeholders** with transparent images
3. update `src` attribute for each image and assign the best quality/size ratio URL

Finally, it will lazy load images to speed up page load time even further.

## Compatibility

Imager is tested against the following mobile and desktop browsers:

- Chrome 33
- Firefox 22
- Opera 12.15
- Edge 12
- Internet Explorer 8, 9, 10 and 11
- Safari 5, 6, 7, 8 and 9
- Mobile Safari 4, 5, 6 and 7
- Android 2, 3 and 4


## Install

npm                            | bower                            | old school
-------------------------------|----------------------------------|------------------------------------------------------------------------------
`npm install --save imager.js` | `bower install --save imager.js` | [download zip file](https://github.com/BBC-News/Imager.js/archive/master.zip)

## Using 2018

Changes below made in 2018 by Ed Horsford

```html
<div style="width: 240px">
    <img class="delayed-image-load" data-src="http://placehold.it/{width}" data-alt="alternative text"></img>
</div>

<script>
    new Imager({ availableWidths: [200, 260, 320, 600] });
</script>
```

This will result in the following HTML output:

```html
<div style="width: 240px">
    <img src="http://placehold.it/260" data-src="http://placehold.it/{width}" alt="alternative text" class="image-replace">
</div>

<script>
    new Imager({ availableWidths: [200, 260, 320, 600] });
</script>
```

`260` has been elected as the best available width (as it is the closest upper size relative to `240` pixels).

### Pixel Ratio / HiDPI / Retina

Imager will automatically multiply the width of the container by the device pixel ratio to request HiDPI images. This can be disabled by setting `multiplyPixelRatio: false`.

Alternately, pixel ratio can be sent separately. Let's say we have generated 4 sizes of images (`200`, `260`, `320` and `600`)
in 3 different pixel ratio flavours (`1`, `1.3` and `2`):

```html
<div style="width: 240px">
    <img class="delayed-image-load" data-src="http://example.com/assets/{width}/imgr{pixel_ratio}.png" data-alt="alternative text"></img>
</div>

<script>
    new Imager({ availableWidths: [200, 260, 320, 600], availablePixelRatios: [1, 1.3, 2], multiplyPixelRatio: false });
</script>
```

The `img[src]` will be computed as following (according to the reported `window.devicePixelRatio` value by the device):

- `http://example.com/assets/260/imgr.png` if no pixel ratio is detected, or advertised as `1`
- `http://example.com/assets/260/imgr-2x.png` if pixel ratio is advertised as `2` (or any value greater than 2)
- `http://example.com/assets/260/imgr-1.3x.png` if pixel ratio is advertised as `1.3`

Head to this [device pixel density test](http://bjango.com/articles/min-device-pixel-ratio/) resource to learn more about the available pixel ratio for your device.

### Interpolating `{width}` value

Imager has the ability to replace `{width}` with a non-numeric value if you
provide the `widthInterpolator` option, which is a function that returns the
string to be injected into the image URL for a given width. This feature allows you to use a human readable name or integrate with third-party image providers.

```html
<div style="width: 240px">
    <img class="delayed-image-load" data-src="http://example.com/assets/imgr-{width}.png" data-alt="alternative text"></img>
</div>

<script>
    new Imager({
        availableWidths: [200, 260, 320, 600],
        widthInterpolator: function(width, pixelRatio) {
          return width + 'x' + (width / 2);
        }
    });
</script>
```

The `img[src]` will be computed as `http://example.com/assets/imgr-260x130.png` instead of `http://example.com/assets/imgr-260.png`.

Alternatively you can define `availableWidths` as an `Object` where the key is the width, and the value is the string to be injected into the image URL.

```html
<div style="width: 240px">
    <img class="delayed-image-load" data-src="http://example.com/assets/imgr-{width}.png" data-alt="alternative text"></img>
</div>

<script>
    new Imager({
        availableWidths: {
            200: 'square',
            260: 'small',
            320: 'medium',
            600: 'large'
        }
    });
</script>
```

The `img[src]` will be computed as `http://example.com/assets/imgr-small.png` instead of `http://example.com/assets/imgr-260.png`.

### Mixing various configurations

You might want to generate HiDPI responsive images. But what if you also include images from another provider which
serves a totally different set of sizes, without pixel ratio?

Here is an example to serve your own images alongside [Flickr images](http://www.flickr.com/).

```html
<div style="width: 240px">
    <img class="delayed-image-load"        data-src="http://placehold.it/{width}" data-alt="alternative text 1"></img>
    <img class="delayed-flickr-image-load" data-src="//farm5.staticflickr.com/4148/4990539658_a38ed4ec6e_{width}.jpg" data-alt="alternative text 2"></img>
</div>

<script>
    var imgrPlaceholder = new Imager('.delayed-image-load', {
        availableWidths: [200, 260, 320, 600]
    });

    var imgrFlickr = new Imager('.delayed-flickr-image-load', {
        availableWidths: {
            150: 't_d',
            500: 'd',
            640: 'z_d'
        }
    });
</script>
```

This will result in the following HTML output:

```html
<div style="width: 240px">
    <img src="http://placehold.it/260" data-src="http://placehold.it/{width}" alt="alternative text 1" class="image-replace">
    <img src="//farm5.staticflickr.com/4148/4990539658_a38ed4ec6e_d.jpg" data-src="//farm5.staticflickr.com/4148/4990539658_a38ed4ec6e_{width}.jpg" alt="alternative text 2" class="image-replace">
</div>
```

### Passing a collection of elements

You might want to pass a NodeList or an array of **placeholder elements** as the first argument rather than a class selector.

```html
<div style="width: 240px">
    <img class="delayed-image-load" data-src="http://placehold.it/{width}" data-alt="alternative text"></img>
</div>

<script>
    var placeholderElems = document.querySelectorAll('.delayed-image-load');
    var imgrPlaceholder = new Imager(placeholderElems, {
        availableWidths: [200, 260, 320, 600]
    });
</script>
```

This will result in the following HTML output:

```html
<div style="width: 240px">
    <img src="http://placehold.it/260" data-src="http://placehold.it/{width}" alt="alternative text" class="image-replace">
</div>
```

# Documentation

Browse Imager public APIs and options – versioned alongside the source code of the project:

- [HTML options](docs/html-api.md)
- [JavaScript options](docs/js-options.md)
- [JavaScript API](docs/js-api.md)

# Demos

Additional and fully working examples lie in the [`demos` folder](demos/).


# They are using it

- [BBC News](http://m.bbc.co.uk/news)
- [BBC Sport](http://m.bbc.co.uk/sport)
- [The Guardian](http://www.theguardian.com/)
- [`x-imager` Web Component](https://github.com/addyosmani/x-imager)
- [Imager.jsx React Component](https://www.npmjs.com/package/imager.jsx)

# Background

This is an experiment in offering developers an interim solution to responsive images based on the [ImageEnhancer](https://gist.github.com/Integralist/6157139) concept researched and developed by the team at BBC News.

At present, support for `srcset` and `PictureFill` are not widespread and the polyfills for these solutions also come with a number of drawbacks.

[Mark McDonnell (@integralist)](http://twitter.com/Integralist) documented the process and rewrote the original code so it could be evolved and improved with the help of the open-source community.

The goal of this project is to automate the process with the help of the [Grunt](http://gruntjs.com/) JavaScript task runner (potentially via `grunt-responsive-images` for image generation based on a source directory).

Much of this work can be repurposed to work with a more standards-based approach once support improves in modern browsers.

For the purposes of maintaining a distinguishment between the ImageEnhancer concept built by BBC News and this project, we're calling it **Imager.js**

**[Read more on BBC Responsive News blog](http://responsivenews.co.uk/post/58244240772/imager-js)**.

# Credits

- [Mark McDonnell](http://twitter.com/integralist)
- [Tom Maslen](http://twitter.com/tmaslen)
- [Thomas Parisot](https://twitter.com/oncletom)
- [Addy Osmani](http://twitter.com/addyosmani)


# Licence

> Copyright 2015 British Broadcasting Corporation
>
> Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at
>
> http://www.apache.org/licenses/LICENSE-2.0
>
> Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.md
## Change Log

### 0.6.0 (2018/09/13 14:30 +00:00)
- Add option to multiply width by device pixel ratio
- Provide new set of default widths
- Add version number information as comment to js
- Remove code to replace divs with imgs - act on imgs directly
- Add option to specify a target attribute instead of src
- Add option to specify a source attribute instead of data-src

### 0.5.0 (2015/01/09 19:30 +00:00)
- [#140](https://github.com/BBC-News/Imager.js/pull/140) Respect alt attribute of placeholder img (@wildlyinaccurate)

### 0.5.0 (2015/01/09 19:30 +00:00)
- [#118](https://github.com/BBC-News/Imager.js/pull/118) Ability to add elements to Imager after initialisation (WebSeed)
- [#113](https://github.com/BBC-News/Imager.js/pull/113) Add Imager.addEvent and Imager.debounce

### 0.3.1 (2014/10/26 15:32 +00:00)
- [#98](https://github.com/BBC-News/Imager.js/issues/98) Firefox 29 requests `undefined` URL in case of lazy loading (@tokkonopapa)
- [#90](https://github.com/BBC-News/Imager.js/issues/90) Fix #89, images not loaded when revealed by resize (@paul-butcher)
- [#89](https://github.com/BBC-News/Imager.js/issues/89) Lazy loading does not load images when revealed by a resize (@paul-butcher)
- [#82](https://github.com/BBC-News/Imager.js/issues/82) JS error: Cannot call method 'replaceChild' of null --- Upon Scroll only (@hi-im-si)
- [#71](https://github.com/BBC-News/Imager.js/issues/71) Issues with resizing of Lazyloaded images (@anthonybruno)
- [#106](https://github.com/BBC-News/Imager.js/pull/106) Imager.ready(fn) (@oncletom)
- [#105](https://github.com/BBC-News/Imager.js/pull/105) Imager.getNaturalWidth(image) (@oncletom)
- [#104](https://github.com/BBC-News/Imager.js/pull/104) Lazyload updates only response images on resize (@oncletom)

### 0.3.0-beta.1 (2014/03/20 12:55 +00:00)
- [#85](https://github.com/BBC-News/Imager.js/pull/85) Fixed reference to js api docs (@omgmog)
- [#83](https://github.com/BBC-News/Imager.js/pull/83) Image width is set as 0 sometimes (@oncletom)
- [#86](https://github.com/BBC-News/Imager.js/pull/86) Containers above max image size load smallest image (@oncletom)

### 0.2.0 (2014/02/24 17:41 +00:00)
- [#51](https://github.com/BBC-News/Imager.js/pull/51) fix typos in README.md (@nikai3d)
- [#50](https://github.com/BBC-News/Imager.js/pull/50) [READY] `imager` already taken in npm registry (@oncletom)
- [#56](https://github.com/BBC-News/Imager.js/pull/56) Add missing </div> end tags (@gunnarbittersmann)
- [#57](https://github.com/BBC-News/Imager.js/pull/57) Update Imager.js: Add support for alt attribute using data-alt (@longprao)
- [#60](https://github.com/BBC-News/Imager.js/pull/60) add offsetTop of each of the element's offsetParents (@marcinwenger)
- [df0acfa](https://github.com/BBC-News/Imager.js/commit/df0acfad0ef8badaeea4d198a339f0c29d940bd9) Merge branch 'fix-loop-order' of https://github.com/oncletom/Imager.js into oncletom-fix-loop-order (@oncletom)
- [5575c4c](https://github.com/BBC-News/Imager.js/commit/5575c4c12ae20a5065ff4aa10af7e981b971b5bf) Merge branch 'fix-404' of https://github.com/oncletom/Imager.js into oncletom-fix-404 (@oncletom)
- [fcdbcea](https://github.com/BBC-News/Imager.js/commit/fcdbcea7d88fc98bc2503505b512de0fe64666e1) Merge branch 'feature-availablePixelRatio' of https://github.com/oncletom/Imager.js into oncletom-feature-availablePixelRatio (@oncletom)
- [#70](https://github.com/BBC-News/Imager.js/pull/70) Sauce Remote Testing (@oncletom)
- [#72](https://github.com/BBC-News/Imager.js/pull/72) [READY] Gecko Compatibility (@oncletom)
- [#74](https://github.com/BBC-News/Imager.js/pull/74) [READY] IE8 Compatible Unit Testing (@oncletom)
- [#73](https://github.com/BBC-News/Imager.js/pull/73) [READY] Internet Explorer 8 Features Compatibility (@oncletom)
- [#75](https://github.com/BBC-News/Imager.js/pull/75) [READY] Browsers compatibility (@oncletom)
- [#76](https://github.com/BBC-News/Imager.js/pull/76) Running SauceLabs on branch, locally otherwise (@oncletom)
- [#66](https://github.com/BBC-News/Imager.js/pull/66) [READY] README demo for defining availableWidths as a function does not work (@oncletom)
- [#68](https://github.com/BBC-News/Imager.js/pull/68) [Fix #67] Add `widthInterpolator` option (@BPScott)
- [e5ddade](https://github.com/BBC-News/Imager.js/commit/e5ddade65371f88bb53ef3d0a5a3fede5ab7c706) Merge remote-tracking branch 'upstream/browsers-compatibility' (@oncletom)
- [#78](https://github.com/BBC-News/Imager.js/pull/78) [READY] Compression and mangle options for uglifyjs when building (@oncletom)

### 0.1.0 (2013/12/08 16:03 +00:00)
- [#49](https://github.com/BBC-News/Imager.js/pull/49) Imager 0.1.0 (@Integralist, @oncletom, @joewillnyc)

### 0.0.2 (2013/11/18 22:28 +00:00)
- [#2](https://github.com/BBC-News/Imager.js/pull/2) imager.js updated (@stephenplusplus)
- [#5](https://github.com/BBC-News/Imager.js/pull/5) Removed all redundant files and created working Demo (@Integralist)
- [#14](https://github.com/BBC-News/Imager.js/pull/14) Automate image generation using Grunt (@Integralist)
- [#17](https://github.com/BBC-News/Imager.js/pull/17) Lazy Loading Images (@Integralist)
- [#29](https://github.com/BBC-News/Imager.js/pull/29) Fixed README typo (@jonlow)
- [#39](https://github.com/BBC-News/Imager.js/pull/39) add main field to package.json (@joewillnyc)
- [#41](https://github.com/BBC-News/Imager.js/pull/41) add support for umd (@joewillnyc)


Imager.js
/*
 
 * Imager.js
 * v0.6.0

 * Original project:
 * https://github.com/BBC-News/Imager.js
 *
 * Forked by Ed Horsford:
 * https://github.com/edwardhorsford/Imager.js

*/

;(function (window, document) {
    'use strict';

    var addEvent = (function () {
        if (document.addEventListener) {
            return function addStandardEventListener(el, eventName, fn) {
                return el.addEventListener(eventName, fn, false);
            };
        }
        else {
            return function addIEEventListener(el, eventName, fn) {
                return el.attachEvent('on' + eventName, fn);
            };
        }
    })();

    var defaultWidths = [50, 100, 150, 200, 250, 300, 400, 500, 600, 700, 800, 900, 1000, 1100, 1200, 1400, 1600, 1800, 2000];

    var getKeys = typeof Object.keys === 'function' ? Object.keys : function (object) {
        var keys = [],
            key;

        for (key in object) {
            keys.push(key);
        }

        return keys;
    };

    var applyEach = function (collection, callbackEach) {
        var i = 0,
            length = collection.length,
            new_collection = [];

        for (; i < length; i++) {
            new_collection[i] = callbackEach(collection[i], i);
        }

        return new_collection;
    };

    var returnFn = function (value) { return value; };
    var noop = function () {};
    var trueFn = function () { return true; };

    var debounce = function (fn, wait) {
        var timeout;
        return function () {
            var context = this, args = arguments;
            var later = function () {
                timeout = null;
                fn.apply(context, args);
            };
            clearTimeout(timeout);
            timeout = setTimeout(later, wait);
        };
    };


    /*
        Construct a new Imager instance, passing an optional configuration object.

        Example usage:

            {
                // Available widths for your images
                availableWidths: [Number],

                // Selector to be used to locate your div placeholders
                selector: '',

                // Class name to give your resizable images
                className: '',

                // If set to true, Imager will update the src attribute of the relevant images
                onResize: Boolean,

                // Toggle the lazy load functionality on or off
                lazyload: Boolean,

                // Used alongside the lazyload feature (helps performance by setting a higher delay)
                scrollDelay: Number
            }

        @param {object} configuration settings
        @return {object} instance of Imager
     */
    var Imager = function (elements, opts) {
        var self = this;
        var doc  = document;

        opts = opts || {};

        if (elements !== undefined) {

            // selector string (not elements)
            if (typeof elements === 'string') {
                opts.selector = elements;
                elements = undefined;
            }

            // 'opts' object (not elements)
            else if (typeof elements.length === 'undefined') {
                opts = elements;
                elements = undefined;
            }
        }

        this.viewportHeight   = doc.documentElement.clientHeight;
        this.selector         = !elements ? (opts.selector || '.delayed-image-load') : null;
        this.className        = opts.className || 'image-replace';
        this.gif              = doc.createElement('img');
        this.gif.src          = 'data:image/gif;base64,R0lGODlhEAAJAIAAAP///wAAACH5BAEAAAAALAAAAAAQAAkAAAIKhI+py+0Po5yUFQA7';
        this.gif.className    = this.className;
        this.gif.alt          = '';
        this.lazyloadOffset   = opts.lazyloadOffset || 0;
        this.scrollDelay      = opts.scrollDelay || 250;
        this.onResize         = opts.hasOwnProperty('onResize') ? opts.onResize : true;
        this.lazyload         = opts.hasOwnProperty('lazyload') ? opts.lazyload : false;
        this.multiplyPixelRatio   = opts.hasOwnProperty('multiplyPixelRatio') ? opts.multiplyPixelRatio : true;
        this.targetAttribute       = opts.targetAttribute || 'src';
        this.sourceAttribute       = opts.sourceAttribute || 'data-src';
        this.scrolled         = false;
        this.availablePixelRatios = opts.availablePixelRatios || [1, 2];
        this.availableWidths  = opts.availableWidths || defaultWidths;
        this.onImagesReplaced = opts.onImagesReplaced || noop;
        this.widthsMap        = {};
        this.refreshPixelRatio();
        this.widthInterpolator = opts.widthInterpolator || returnFn;

        // Needed as IE8 adds a default `width`/`height` attribute…
        this.gif.removeAttribute('height');
        this.gif.removeAttribute('width');

        if (typeof this.availableWidths !== 'function') {
          if (typeof this.availableWidths.length === 'number') {
            this.widthsMap = Imager.createWidthsMap(this.availableWidths, this.widthInterpolator, this.devicePixelRatio);
          }
          else {
            this.widthsMap = this.availableWidths;
            this.availableWidths = getKeys(this.availableWidths);
          }

          this.availableWidths = this.availableWidths.sort(function (a, b) {
            return a - b;
          });
        }

        this.divs = [];
        this.add(elements || this.selector);
        this.ready(opts.onReady);

        setTimeout(function () {
            self.init();
        }, 0);
    };

    Imager.prototype.add = function (elementsOrSelector) {

        elementsOrSelector = elementsOrSelector || this.selector;
        var elements = typeof elementsOrSelector === 'string' ?
            document.querySelectorAll(elementsOrSelector) : // Selector
            elementsOrSelector; // Elements (NodeList or array of Nodes)

        if (elements && elements.length) {
            var additional = applyEach(elements, returnFn);
            // Not sure this check is needed
            if (this.initialized) {
                this.checkImagesNeedReplacing(elements);
            }
            this.divs = this.divs.concat(additional);
        }
    };

    Imager.prototype.scrollCheck = function () {
        var self = this;
        var offscreenImageCount = 0;
        var elements = [];

        if (this.scrolled) {
            // collects a subset of not-yet-responsive images and not offscreen anymore
            applyEach(this.divs, function (element) {
                if (self.isPlaceholder(element)) {
                    ++offscreenImageCount;

                    if (self.isThisElementOnScreen(element)) {
                        elements.push(element);
                    }
                }
            });

            if (offscreenImageCount === 0) {
                window.clearInterval(self.interval);
            }

            // This check might not be needed
            if (this.initialized) {
                this.checkImagesNeedReplacing(elements);
            }

            this.scrolled = false;
        }
    };

    Imager.prototype.init = function () {
        var self = this;

        this.initialized = true;
        var filterFn = trueFn;

        if (this.lazyload) {
            this.registerScrollEvent();

            this.scrolled = true;
            self.scrollCheck();

            filterFn = function (element) {
                return self.isPlaceholder(element) === false;
            };
        }
        else {
            this.checkImagesNeedReplacing(this.divs);
        }

        if (this.onResize) {
            this.registerResizeEvent(filterFn);
        }

        this.onReady();
    };

    /**
     * Executes a function when Imager is ready to work
     * It acts as a convenient/shortcut for `new Imager({ onReady: fn })`
     *
     * @since 0.3.1
     * @param {Function} fn
     */
    Imager.prototype.ready = function (fn) {
        this.onReady = fn || noop;
    };

    // Imager.prototype.createGif = function (element) {
    //     // if the element is already a responsive image then we don't replace it again
    //     if (element.className.match(new RegExp('(^| )' + this.className + '( |$)'))) {
    //         return element;
    //     }

    //     var elementClassName = element.getAttribute('data-class');
    //     var elementWidth = element.getAttribute('data-width');
    //     var gif = this.gif.cloneNode(false);

    //     if (elementWidth) {
    //       gif.width = elementWidth;
    //       gif.setAttribute('data-width', elementWidth);
    //     }

    //     gif.className = (elementClassName ? elementClassName + ' ' : '') + this.className;
    //     gif.setAttribute('data-src', element.getAttribute('data-src'));
    //     gif.setAttribute('alt', element.getAttribute('data-alt') || element.alt || this.gif.alt);

    //     element.parentNode.replaceChild(gif, element);

    //     return gif;
    // };

    // Imager.prototype.changeDivsToEmptyImages = function (elements) {
    //     var self = this;

    //     applyEach(elements, function (element, i) {
    //         elements[i] = self.createGif(element);
    //     });

    //     if (this.initialized) {
    //         this.checkImagesNeedReplacing(elements);
    //     }
    // };

    /**
     * Indicates if an element is an Imager placeholder
     *
     * @since 1.3.1
     * @param {HTMLImageElement} element
     * @returns {boolean}
     */
    Imager.prototype.isPlaceholder = function (element) {
        return element.src === this.gif.src;
    };

    /**
     * Returns true if an element is located within a screen offset.
     *
     * @param {HTMLElement} element
     * @returns {boolean}
     */
    Imager.prototype.isThisElementOnScreen = function (element) {
        // document.body.scrollTop was working in Chrome but didn't work on Firefox, so had to resort to window.pageYOffset
        // but can't fallback to document.body.scrollTop as that doesn't work in IE with a doctype (?) so have to use document.documentElement.scrollTop
        var elementOffsetTop = 0;
        var offset = Imager.getPageOffset() + this.lazyloadOffset;

        if (element.offsetParent) {
            do {
                elementOffsetTop += element.offsetTop;
            }
            while (element = element.offsetParent);
        }

        return elementOffsetTop < (this.viewportHeight + offset);
    };

    Imager.prototype.checkImagesNeedReplacing = function (images, filterFn) {
        var self = this;
        filterFn = filterFn || trueFn;

        if (!this.isResizing) {
            this.isResizing = true;
            this.refreshPixelRatio();

            applyEach(images, function (image) {
                if (filterFn(image)) {
                    self.replaceImagesBasedOnScreenDimensions(image);
                }
            });

            this.isResizing = false;
            this.onImagesReplaced(images);
        }
    };

    /**
     * Upgrades an image from an empty placeholder to a fully sourced image element
     *
     * @param {HTMLImageElement} image
     */
    Imager.prototype.replaceImagesBasedOnScreenDimensions = function (image) {
        var computedWidth, naturalWidth;

    naturalWidth = Imager.getNaturalWidth(image);
        computedWidth = typeof this.availableWidths === 'function' ? this.availableWidths(image)
                                                                   : this.determineAppropriateResolution(image);

        image.width = computedWidth;

        if (!this.isPlaceholder(image) && computedWidth <= naturalWidth) {
            return;
        }

        var src = this.changeImageSrcToUseNewImageDimensions(image.getAttribute(this.sourceAttribute), computedWidth);
        image.setAttribute(this.targetAttribute, src);
        var classSelector = this.selector;
        if (classSelector.startsWith(".")){
            classSelector = classSelector.substring(1);
        }
        if (image.classList.contains(classSelector)){
            image.classList.remove(classSelector);
        };
        if (!image.classList.contains(this.className)){
            image.classList.add(this.className);
        };
        image.removeAttribute('width');
        image.removeAttribute('height');
    };

    Imager.prototype.determineAppropriateResolution = function (image) {
        var targetWidth = image.getAttribute('data-width') || image.parentNode.clientWidth;
        if (this.multiplyPixelRatio){
            var pixelRatio = Imager.getPixelRatio();
            targetWidth = targetWidth * pixelRatio;
        }
        console.log("target width is", targetWidth)
      return Imager.getClosestValue(targetWidth, this.availableWidths);
    };

    /**
     * Updates the device pixel ratio value used by Imager
     *
     * It is performed before each replacement loop, in case a user zoomed in/out
     * and thus updated the `window.devicePixelRatio` value.
     *
     * @api
     * @since 1.0.1
     */
    Imager.prototype.refreshPixelRatio = function refreshPixelRatio() {
        this.devicePixelRatio = Imager.getClosestValue(Imager.getPixelRatio(), this.availablePixelRatios);
    };

    Imager.prototype.changeImageSrcToUseNewImageDimensions = function (src, selectedWidth) {
        return src
            .replace(/{width}/g, Imager.transforms.width(selectedWidth, this.widthsMap))
            .replace(/{pixel_ratio}/g, Imager.transforms.pixelRatio(this.devicePixelRatio));
    };

    Imager.getPixelRatio = function getPixelRatio(context) {
        return (context || window)['devicePixelRatio'] || 1;
    };

    Imager.createWidthsMap = function createWidthsMap (widths, interpolator, pixelRatio) {
        var map = {},
            i   = widths.length;

        while (i--) {
            map[widths[i]] = interpolator(widths[i], pixelRatio);
        }

        return map;
    };

    Imager.transforms = {
        pixelRatio: function (value) {
            return value === 1 ? '' : '-' + value + 'x';
        },
        width: function (width, map) {
            return map[width] || width;
        }
    };

    /**
     * Returns the closest upper value.
     *
     * ```js
     * var candidates = [1, 1.5, 2];
     *
     * Imager.getClosestValue(0.8, candidates); // -> 1
     * Imager.getClosestValue(1, candidates); // -> 1
     * Imager.getClosestValue(1.3, candidates); // -> 1.5
     * Imager.getClosestValue(3, candidates); // -> 2
     * ```
     *
     * @api
     * @since 1.0.1
     * @param {Number} baseValue
     * @param {Array.<Number>} candidates
     * @returns {Number}
     */
    Imager.getClosestValue = function getClosestValue(baseValue, candidates) {
        var i             = candidates.length,
            selectedWidth = candidates[i - 1];

        baseValue = parseFloat(baseValue);

        while (i--) {
            if (baseValue <= candidates[i]) {
                selectedWidth = candidates[i];
            }
        }

        return selectedWidth;
    };

    Imager.prototype.registerResizeEvent = function (filterFn) {
        var self = this;

        addEvent(window, 'resize', debounce(function () {
            self.checkImagesNeedReplacing(self.divs, filterFn);
        }, 100));
    };

    Imager.prototype.registerScrollEvent = function () {
        var self = this;

        this.scrolled = false;

        this.interval = window.setInterval(function () {
            self.scrollCheck();
        }, self.scrollDelay);

        addEvent(window, 'scroll', function () {
            self.scrolled = true;
        });

        addEvent(window, 'resize', function () {
            self.viewportHeight = document.documentElement.clientHeight;
            self.scrolled = true;
        });
    };

    Imager.getPageOffsetGenerator = function getPageVerticalOffset(testCase) {
        if (testCase) {
            return function () { return window.pageYOffset; };
        }
        else {
            return function () { return document.documentElement.scrollTop; };
        }
    };

    /**
     * Returns the naturalWidth of an image element.
     *
     * @since 1.3.1
     * @param {HTMLImageElement} image
     * @return {Number} Image width in pixels
     */
    Imager.getNaturalWidth = (function () {
        if ('naturalWidth' in (new Image())) {
            return function (image) {
                return image.naturalWidth;
            };
        }
        // non-HTML5 browsers workaround
        return function (image) {
            var imageCopy = document.createElement('img');
            imageCopy.src = image.targetAttribute;
            return imageCopy.width;
        };
    })();

    // This form is used because it seems impossible to stub `window.pageYOffset`
    Imager.getPageOffset = Imager.getPageOffsetGenerator(Object.prototype.hasOwnProperty.call(window, 'pageYOffset'));

    // Exporting for testing and convenience purpose
    Imager.applyEach = applyEach;
    Imager.addEvent = addEvent;
    Imager.debounce = debounce;

    /* jshint ignore:start */
    if (typeof module === 'object' && typeof module.exports === 'object') {
        // CommonJS, just export
        module.exports = exports = Imager;
    } else if (typeof define === 'function' && define.amd) {
        // AMD support
        define(function () { return Imager; });
    } else if (typeof window === 'object') {
        // If no AMD and we are in the browser, attach to window
        window.Imager = Imager;
    }
    /* jshint ignore:end */

}(window, document));
